### ISSUE TITLE
Configure Docker

### ISSUE THREAD

**Issue by felipevolpone** [openhatch/oh-mainline#1744]
Hey guys, I was thinking about create a Docker image to the project. This can be a nice way to have the project enviroment in a easily way. What do you think?

**Comment by andrefreitas** [comment]
Docker is actually the simplest way of delivering applications in a production or development environment. So yes, it is a good idea :)

I will send a pull request with the Dockerfile soon.

**Comment by felipevolpone** [comment]
@andrefreitas Are you sure? I can do this too, so tell me if you're too busy to do it.

**Comment by paulproteus** [comment]
Hi all,

Thanks for your enthusiasm here!

For what it's worth, the codebase bundles all its dependencies, and works
on Mac OS & Windows, so it's not clear to me how Docker would help. But I'm
curious to hear details.

Thanks!

Asheesh.

**Comment by andrefreitas** [comment]
@paulproteus the key advantages about having a Docker container are:
- [1] You don't need to install the Python Packages dependencies everytime you deploy the application, since in the Dockerfile they are resolved in the build process. You save time and, the cost of deploying, is just starting the container;
- [2] You just need to make sure the Host only have Docker installed;
- [3] You can make the container start automatically when the system is booting so you don't need to configure a service in Systemd, Upstart, etc for this purpose.
- [4] You can use tools to orchestrate (e.g. Scale) such as Rancher, Kubernetes, Tutum.

Bottom line is, this is not a new feature, is just "sugar" for Devops.

@felipevolpone let's collaborate :)

- André Freitas

**Comment by andrefreitas** [comment]
@paulproteus can you describe the steps to setup the environment and have a basic functional version of openhatch? I am doing:

```
pip install -r requirements.txt
python manage.py migrate
python manage.py runserver 0.0.0.0:80
```

And getting the error:
```
UnknownMigration: Migration 'httpproxy:0001_initial' probably doesn't exist.
Traceback (most recent call last):
  File "/code/vendor/packages/south/south/migration/base.py", line 310, in migration
    migration = __import__(full_name, {}, {}, ['Migration'])
  File "/code/vendor/packages/django-http-proxy/httpproxy/migrations/0001_initial.py", line 4, in <module>
    from django.db import models, migrations
ImportError: cannot import name migrations
```

**Comment by felipevolpone** [comment]
@andrefreitas you can check this issue https://github.com/openhatch/oh-mainline/issues/1734 to solve this problem. 

You have to checkout to the last safety commit, run the command and checkout to master.

**Comment by andrefreitas** [comment]
Hi, I just sent the PR #1745. Thanks for your help @felipevolpone  :)

**Comment by paulproteus** [comment]
Thanks for replying!

> [1] You don't need to install the Python Packages dependencies everytime
you deploy the application, since in the Dockerfile they are resolved in
the build process. You save time and, the cost of deploying, is just
starting the container;

We already have this; have you looked at our manage.py?

> [2] You just need to make sure the Host only have Docker installed;

For us, we don't even need Docker. You need Python installed. See
openhatch.readthedocs.org.

> [3] You can make the container start automatically when the system is
booting so you don't need to configure a service in Systemd, Upstart, etc
for this purpose.

We already have a uwsgi configuration that can work for this, if needed.
But we use Heroku now anyway.

> [4] You can use tools to orchestrate (e.g. Scale) such as Rancher,
Kubernetes, Tutum.

We use Heroku so that doesn't affect us.

**Comment by felipevolpone** [comment]
@paulproteus @andrefreitas Actually when I proposed create a Docker configuration it was just to make easier to configure the enviroment to developers that want contribute to the project. In my case, I tried to run the pip install and I got a conflict in the version of psycopg, because I already have an another version of it in my machine. 

Anyway, if the maintainers don't see any value in this issue, I'll try to get another issue. Thank you guys

**Comment by ehashman** [comment]
Hi @felipevolpone, I am confused as to why you are running pip install---can you please walk me through the steps of how you are trying to develop the site? [Here's our installation documentation](http://openhatch.readthedocs.org/en/latest/getting_started/installation.html) for reference. All of oh-mainline's dependencies are bundled in the source code and you do not need pip to install dependencies. Perhaps you are referring to oh-bugimporters?

I said some things on IRC on the topic of why I don't think Docker is appropriate for this repo (oh-mainline) but I'll summarize/reiterate them here:

- Docker adds a ton of complexity and is not crossplatform: we support both Linux, Windows, and Mac OS X development out of the box, and Docker doesn't help us do that
- Our project is designed to be as beginner-friendly as possible, easy to install by people who have never coded before; Docker is advanced Linux containerization tool for experienced developers that may confuse newbies
- There is only one instance of the OpenHatch site; Docker is more useful when one needs to continuously deploy many containers
- Docker solves a problem we don't have: when everything you need to deploy is included in the source code, why create a custom clean environment to satisfy dependencies that have already been met? Vagrant would be another possible suggestion here, but again, it's not actually solving a problem we have.

I'd be really happy to help you out with the install process if you're having difficulty, @felipevolpone; please file another issue if you are having problems with that or hop on #openhatch on FreeNode and ping me. I'm going to close this specific issue as "not applicable."

**Comment by felipevolpone** [comment]
Hey @ehashman, You're totally right, I ran pip install -r requirements.txt 'cause I saw the requirements file and didn't followed the instructions in the documentation. Sorry for this.
